### PR TITLE
Authentication Integration

### LINKED ISSUE #6: Log in/Register
As a user, I want to log in, so that I can see the machine data relevant to me.

**Acceptance Criteria**

 - Be able to register
 - Be able to log in
 - Using an incorrect username/password prevents getting to the profile/data pages
 - You must be logged in to see the profile/data pages
- Admin Accounts?

**Is your feature request related to a problem? Please describe.**
You should only be able to see what's relevant to you, and users should have the confidence that others can't see their data. Furthermore we need a way of authenticating users.

**Describe the solution you'd like**
A user must be able to log in using username and password. This username and password combination should not be hard-coded (i.e. actual combinations must be stored in a sort of database and checked).

**Describe alternatives you've considered**
No alternatives considered.

**Additional context**
There is a google plugin of sorts (can't remember the name), which will help speed up set up.

### PR DESCRIPTION
**Pull request by Deagler**
## GitHub Issue Solved:

closes #6 

## Current behaviour

There was no login/register

## Changed behaviour

- Added FirebaseUI to the frontend (using Google OAuth for now for speed of development)
- Added a useUserContext custom hook to get firebase user credentials
- Added a user detail + log out button to the navigation bar
- Backend now parses the user id from the Authorization header. 
- App auto redirects to login page if you aren't logged in yet.

**This means you need to set LOCAL_MOCK_USER in api/functions/.env** (only if you want to use the GraphQL Playground) The current user ID there is my user id.

## PR checklist

Remember to check the following

 - [x] Tag specific people to review your PR
 - [x] Update the ReadMe with any new run instructions
 - [x] Closes the associated issue (if applicable)

### Demo
image

image

### CODE REVIEW

**Review by FrankWangma — CHANGES_REQUESTED** [review]
Everything looks good to me for the most part.
Just one part which isn't really an issue, but more of a QoL change: 
- The logout button immediately logs the user out. I was wondeing if we should have like a prompt, to ask whether or not the user wants to log out just in case they accidentally clicked the button

File: app/src/components/Heading.tsx [hunk 1]
```diff
@@ -10,30 +11,60 @@ import {
...
+                        text: "Okay",
+                        handler: () => {
+                            firebaseAuth.signOut();
+                        },
+                    },
+                ]}
+            />
+            <IonToolbar color="dark">
+                <div className="w-full flex justify-between items-center">
+                    <IonBackButton />
```
Comment (merw957) [inline app/src/components/Heading.tsx:52]: The back button isn't showing up for me. I think you need to put the <IonButton> thing around this again for it to work.

File: app/src/components/Heading.tsx [hunk 2]
```diff
@@ -10,30 +11,60 @@ import {
...
+                    <div className="block">
+                        <IonTitle className="text-center text-2xl font-heading">
```
Comment (merw957) [inline app/src/components/Heading.tsx:54]: Also need to put the font back on this title

**Review by FrankWangma — APPROVED** [review]
Looking good to me now

**Pull request merged**